OpenQASM read capability

sarsid opened the pull request:
**Description**
The OpenQASM 2.0 standard is described [here](https://github.com/Qiskit/openqasm/blob/master/spec/qasm2.rst). 
The purpose of this PR is to create a full implementation of a OpenQASM parser that reads 
a .qasm circuit into a QubitCircuit. This function lives in qutip/qip/qasm.py. There are a number of 
internal functions that enable this conversion but the only api-exposed function is supposed to be 
`read_qasm` which takes in a .qasm file and returns a fully-formed QubitCircuit object. Currently, the heft of the work is done with processing user-defined gates and adding them to the circuit (as well as measurements, but they need to have #1274 on measurements merged before I can enable that functionality). I still need to add features pertaining to initialization of qubits as well as classical control. 



**Related issues or PRs**
There was a related PR #1140 which purportedly worked with OpenQASM 1.0 but this one deviates from it largely while adding additional features as well as the ability to process user-defined gates and a comparatively robust parser. 
**Changelog**
Added read_qasm function to read .qasm files into QubitCircuit

nathanshammah commented:
@sarsid thank you for the detailed description of the PR. Some tests are failing at [` TestQubitCircuit.test_user_gate`](https://travis-ci.org/github/qutip/qutip/jobs/699177830#L3017), "NameError: name 'isfunction' is not defined". 

This is still in draft mode, however I wonder whether you think it makes sense to provide the user the option to choose from OpenQasm 2.0 and Qasm, with the same function, e.g., `read_qasm`, but with an attribute. 

This is a draft, but please remember to add Numpy docstrings to functions. 

It would be great to have read/write capabilities with QASM, looking forward to this PR.

ajgpitch commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,400 @@
+from math import pi
+
+
+class QASMGate:
```
I prefer the QasmGate style name

ajgpitch commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,400 @@
+        self.gates_inside = []
+
+
+def _tokenize(tok_commands):
```
maybe `token_cmds` is a better abbr.

ajgpitch commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,400 @@
+    return gate_args, gate_regs
+
+
+class QASMProcessor:
```
`QasmProcessor`

ajgpitch commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,400 @@
+
+    def _regs_processor(self, regs, reg_type):
+        '''
+        Processes register tokens mapping them to the QubitCircuit index
+        of the respective registers.
```
Comments should be read like a command to the code that follows. I don't where I read this, but I am sure it's in [pep8](https://www.python.org/dev/peps/pep-0008/#comments) somewhere.
(I want to do that thing that @nathanshammah showed us where we make suggested changes. Cant see the option)
`Process the register tokens: map them to the QubitCircuit index of the respective registers.`

ajgpitch commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,400 @@
+
+    def _initialize_pass(self):
+        '''
+        Passes through the tokenized commands to create QASMGate objects for
```
as previous

sarsid commented:
Update : Most of the functionality is now there, except `measure` and conditional statements, I hope to add those after we are done merging the measurement PR. @nathanshammah I was wondering if writing to QASM should be a separate PR or the same one ?

nathanshammah commented:
Thank you Sidhant. I’d split it up.

On Sat, 27 Jun 2020 at 06:07, Sidhant Saraogi <notifications@github.com>
wrote:

> Update : Most of the functionality is now there, except measure and
> conditional statements, I hope to add those after we are done merging the
> measurement PR. @nathanshammah <https://github.com/nathanshammah> I was
> wondering if writing to QASM should be a separate PR or the same one ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qutip/qutip/pull/1291#issuecomment-650486135>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADPF67AEYTRJEAW23HB3JLDRYVV7LANCNFSM4OAG5ZOA>
> .
>
-- 
Dr. Nathan Shammah
www.nathanshammah.com

sarsid commented:
> @sarsid thank you for the detailed description of the PR. Some tests are failing at [` TestQubitCircuit.test_user_gate`](https://travis-ci.org/github/qutip/qutip/jobs/699177830#L3017), "NameError: name 'isfunction' is not defined".
> 
> This is still in draft mode, however I wonder whether you think it makes sense to provide the user the option to choose from OpenQasm 2.0 and Qasm, with the same function, e.g., `read_qasm`, but with an attribute.
> 
> This is a draft, but please remember to add Numpy docstrings to functions.
> 
> It would be great to have read/write capabilities with QASM, looking forward to this PR.

@nathanshammah  I was wondering if functions only meant for internal class use should have returns and parameters specified for internal functions?

BoxiLi commented:
> I was wondering if functions only meant for internal class use should have returns and parameters specified?

It definitely helps whoever reading the code and trying to figure out what it is doing. It's always good to have more doc I would say, so long as it is correct.

sarsid commented:
I am trying to add some `.qasm` files to the folder `qutip/qutip/tests/qasm_files/` to use during test time. I thought it would be enough to add them to the `PACKAGE_DATA` variable in `setup.py` but that does not seem to work. Any ideas on how to make sure this folder is available during testing time.

nathanshammah commented:
would adding it to the manifest (MANIFEST.in) help?

BoxiLi commented:
I think it should be possible to do this within the pytest framework without touching the package `setup.py` or `MANIFEST`. [`pytest.fixture`](https://docs.pytest.org/en/stable/fixture.html#sharing-test-data) seems a natural choice for loading data. For file accessibility, [this](https://stackoverflow.com/questions/46019170/how-do-you-properly-integrate-unit-tests-for-file-parsing-with-pytest) might help. Porbably @jakelishman knows more about the natural way of doing this with pytest?

jakelishman commented on qutip/tests/test_qasm.py:
```diff
@@ -0,0 +1,117 @@
+                 "QASM: File does not contain QASM 2.0 header")])
+def test_qasm_errors(filename, error, error_regex):
+    filepath = Path(__file__).parent
+    np.testing.assert_raises_regex(error, error_regex, read_qasm,
+                                   "{}/qasm_files/{}".format(filepath, filename))
```
It's a little easier to read what's going on if you spell this like
```python
path = Path(__file__).parent / 'qasm_files' / filename
with pytest.raises(error) as exc_info:
    read_qasm(path)
assert error_message in str(exc_info.value)
```

jakelishman commented:
As I understand it, the only way to avoid needing to touch `MANIFEST.in` is to not have files distributed with the test suite (or to have the temporary files by `*.py`, since `setup.py` automatically recognises them).  That's not really a pytest thing, it's a `setuptools` one.

I don't think there's a problem distributing files as part of the tests.  You can add a line which says
```
recursive-include qutip/tests/qasm_files/ *.qasm
```
to `MANIFEST.in`, and that should hopefully fix the problem.  You won't have the problem locally, because you'll be running `./setup.py develop`, which "installs" in-place, so you don't notice the fact that you've missed files from the manifest.  In the tests, it runs `./setup.py install`, which does the copy, missing files which aren't in the manifest.

Boxi: I'm not certain what more you'd like to do with fixtures here. Fixtures are good for setup and tear down, parametrisation and sharing resources (like network connections), but I'm not certain what further benefit you get for a test function which is testing whether it can open a known file, read it, and construct the expected output.  I think Sidhant is already doing some nice parametrisation of his tests, from what I see.

Sidhant: I added a comment on the exception test, since pytest gives us an easier-to-read context manager you can use.  It's also typically better practice to handle file paths like I did in that (using the overloaded `/` operator with `pathlib.Path`, or `os.path.join`) rather than manually putting in slashes.  Windows _usually_ will do the right thing now with slashes, but it's better to let Python handle it for us.

BoxiLi commented:
The conflict is just because that trailing white space PR got merged. You can update from qutip master and resolve it locally. No worries.

sarsid commented:
> The conflict is just because that trailing white space PR got merged. You can update from qutip master and resolve it locally. No worries.

I don't think it is that @BoxiLi, I need to change the manifest as per @jakelishman's comments to include the qasm_files folder. Is there a different way to do this?

sarsid commented:
> As I understand it, the only way to avoid needing to touch `MANIFEST.in` is to not have files distributed with the test suite (or to have the temporary files by `*.py`, since `setup.py` automatically recognises them). That's not really a pytest thing, it's a `setuptools` one.
> 
> I don't think there's a problem distributing files as part of the tests. You can add a line which says
> 
> ```
> recursive-include qutip/tests/qasm_files *.qasm
> ```
> 
> to `MANIFEST.in`, and that should hopefully fix the problem. You won't have the problem locally, because you'll be running `./setup.py develop`, which "installs" in-place, so you don't notice the fact that you've missed files from the manifest. In the tests, it runs `./setup.py install`, which does the copy, missing files which aren't in the manifest.
> 
> Boxi: I'm not certain what more you'd like to do with fixtures here. Fixtures are good for setup and tear down, parametrisation and sharing resources (like network connections), but I'm not certain what further benefit you get for a test function which is testing whether it can open a known file, read it, and construct the expected output. I think Sidhant is already doing some nice parametrisation of his tests, from what I see.
> 
> Sidhant: I added a comment on the exception test, since pytest gives us an easier-to-read context manager you can use. It's also typically better practice to handle file paths like I did in that (using the overloaded `/` operator with `pathlib.Path`, or `os.path.join`) rather than manually putting in slashes. Windows _usually_ will do the right thing now with slashes, but it's better to let Python handle it for us.

Thanks for this information, it seems to work ! I was wondering if it's still needed (at all) in `PACKAGE_DATA` in setup.py?

BoxiLi commented:
After the two qasm PRs get merged, we can also consider make `circuit` a folder aka a submodule. The old API of QubitCircuit will not change: `from qutip.qip.circuit import QubitCircuit`. But we can then have `from qutip.qip.circuit import qasm` without making `circuit.py` three thousand lines long. Like in the `device` folder right now. I find this categorization pretty nice, let's discuss in the meeting ;)

sarsid commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,735 @@
+                    self.num_cbits += num_regs
+                else:
+                    raise SyntaxError("QASM: incorrect bracket formatting")
+            elif command[0] in ["barrier", "include", "reset"]:
```
Yep, that is what reset is meant to do? I did not think to accommodate it since it seems like a scarcely used feature and did not merit a separate kind of processing with QuTiP. What do you think? Maybe a warning is appropriate here? Or an error?

BoxiLi replied:
If `reset` is not supported, i.e. qasm circuit with `reset` will be wrong in QuTiP, it deserves a `NotImplementedError`.

sarsid commented on qutip/qip/qasm.py:
```diff
@@ -0,0 +1,735 @@
+                    qc.add_measurement("M", targets=[regs[0]],
+                                       classical_store=regs[1])
+            elif command[0] == "if":
+                warnings.warn(("Information about individual registers"
+                              " is not preserved in QubitCircuit"))
```
The same is also true for qubit registers! The only difference is that in that case you can't control on multiple qubits explicitly (without it being implicit in a gate).

BoxiLi replied:
I see. that makes sense. Is the converted circuit still the correct one in terms of the final result?

sarsid replied:
Yes, that is the case !

BoxiLi commented:
@sarsid Could you also add an option where I can give a string read from a qasm file instead of the file name? I have a string

We could, for example, change the parameter name from `file` to something more general like `qasm` and add an optional parameter giving the type `"qasm_file"` or `"qasm_str"`. Probably these are not the best name, but something like this would be great.

sarsid commented:
> @sarsid Could you also add an option where I can give a string read from a qasm file instead of the file name? I have a string created elsewhere and now I have to save it to a file just to read it again.
> 
> We could, for example, change the parameter name from `file` to something more general like `qasm` and add an optional parameter giving the type `"qasm_file"` or `"qasm_str"`. Probably these are not the best name, but something like this would be great.

Added! Also fixed all the other suggestions !

BoxiLi commented:
This looks pretty good to me. I'll merge if there are no other suggestions.

nathanshammah commented:
This is great, congrats.
